File: hrms/hr/report/attendance_day_audit/attendance_day_audit.js
// Copyright (c) 2026, Nastyworldwide-Dev and contributors
// License: GNU General Public License v3. See license.txt

frappe.query_reports["Attendance Day Audit"] = {
	filters: [
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.add_days(frappe.datetime.get_today(), -14),
			reqd: 1,
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1,
		},
		{ fieldname: "employee", label: __("Employee"), fieldtype: "Link", options: "Employee" },
		{
			fieldname: "verdict",
			label: __("Verdict"),
			fieldtype: "Select",
			options: [
				"All",
				"punches-skip-stamped",
				"punches-linked-to-cancelled-row",
				"unread-punches",
				"half-day-one-punch",
				"shift-mismatch",
				"before-process-attendance-after",
				"after-last-sync",
				"row-manual",
				"row-mirrored",
				"row-leave",
				"punches-rejected",
				"punches-mirrored",
				"marked",
				"no-punches",
			],
			default: "All",
		},
	],

	onload(report) {
		// Server side is System Manager only; dry run first, then an explicit confirm.
		report.page.add_inner_button(__("Repair Skip Stamps and Dead Links"), () =>
			repair(report)
		);
	},
};

function repair(report) {
	const args = {
		from_date: report.get_filter_value("from_date"),
		to_date: report.get_filter_value("to_date"),
	};
	frappe.call({
		method: "hrms.utils.attendance_day_audit.repair_attendance_days",
		args: { ...args, dry_run: 1 },
		freeze: true,
		callback: ({ message }) => {
			const plan = message.plan || [];
			if (!plan.length) {
				frappe.msgprint(__("Nothing to repair in this window."));
				return;
			}
			const esc = frappe.utils.escape_html;
			const rows = plan
				.map(
					(p) =>
						`<tr><td>${esc(p.employee)}</td><td>${esc(p.date)}</td><td>${esc(
							p.action
						)}</td><td>${p.punches.length}</td></tr>`
				)
				.join("");
			frappe.msgprint({
				title: __("Dry run: {0} employee-day(s)", [plan.length]),
				message: `<table class="table table-bordered table-sm"><thead><tr><th>${__(
					"Employee"
				)}</th><th>${__("Date")}</th><th>${__("Action")}</th><th>${__(
					"Punches"
				)}</th></tr></thead><tbody>${rows}</tbody></table>`,
				wide: true,
			});
			frappe.confirm(
				__(
					"Clear the old skip stamps / dead links on {0} day(s) and queue the hourly job now? Attendance rows themselves are not changed.",
					[plan.length]
				),
				() =>
					frappe.call({
						method: "hrms.utils.attendance_day_audit.repair_attendance_days",
						args: { ...args, dry_run: 0, remark_now: 1 },
						freeze: true,
						callback: ({ message: done }) => {
							frappe.show_alert({
								message: __("Cleared {0} punch(es); hourly job queued: {1}", [
									done.touched,
									done.remark_queued ? __("yes") : __("no"),
								]),
								indicator: "green",
							});
							report.refresh();
						},
					})
			);
		},
	});
}

File: hrms/hr/report/attendance_day_audit/attendance_day_audit.py
"""Attendance Day Audit — script report.

One row per employee-day: the punches, the attendance row, and the VERDICT —
why the day reads the way it does, in the words of the mechanism
(hrms/utils/attendance_day_audit.py judge_day) — plus what the repair would do.
Built for 7 September 2026, when staff had punched and the calendar said
Absent, Half Day or nothing, each for a different reason.
"""

from __future__ import annotations

import logging

import frappe
from frappe import _
from frappe.utils import add_days, getdate

from hrms.utils import attendance_day_audit

logger = logging.getLogger(__name__)

VERDICT_LABELS = {
	"marked": "Marked from its punches",
	"half-day-one-punch": "Half Day from one punch",
	"unread-punches": "Unread punches — waiting for the hourly job",
	"punches-skip-stamped": "Punches skip-stamped (old failure)",
	"punches-linked-to-cancelled-row": "Punches linked to a cancelled row",
	"punches-mirrored": "Punches are mirrored rows",
	"punches-rejected": "Punches rejected",
	"punch-without-shift": "Punch resolved to no shift",
	"shift-mismatch": "Punch shift ≠ attendance shift",
	"shift-auto-attendance-off": "Shift has auto attendance off",
	"before-process-attendance-after": "Before Process Attendance After",
	"after-last-sync": "Job has not reached this shift yet",
	"row-financially-locked": "Payroll / claim depends on this day — HR corrects by hand",
	"row-manual": "Manual attendance row",
	"row-mirrored": "Mirrored attendance row",
	"row-leave": "Leave record",
	"no-punches": "No punches",
}


def execute(filters=None):
	filters = frappe._dict(filters or {})
	if not filters.get("to_date"):
		filters["to_date"] = frappe.utils.today()
	if not filters.get("from_date"):
		filters["from_date"] = add_days(getdate(filters["to_date"]), -14)
	logger.info("[attendance_day_audit] execute filters=%s", dict(filters))
	days = attendance_day_audit.collect(filters.from_date, filters.to_date, filters.get("employee"))["days"]
	wanted = filters.get("verdict")
	if wanted and wanted != "All":
		days = [d for d in days if d["verdict"] == wanted]
	rows = [{**d, "verdict_label": _(VERDICT_LABELS.get(d["verdict"], d["verdict"]))} for d in days]
	repairs = sum(1 for d in days if d.get("repair"))
	message = _("{0} employee-days · {1} repairable by clearing old skip stamps / dead links").format(
		len(days), repairs
	)
	return _columns(), rows, message


def _columns():
	def col(fieldname, label, fieldtype="Data", width=120, options=None):
		column = {"fieldname": fieldname, "label": _(label), "fieldtype": fieldtype, "width": width}
		if options:
			column["options"] = options
		return column

	return [
		col("employee", "Employee", "Link", 130, "Employee"),
		col("employee_name", "Name", width=160),
		col("date", "Date", "Date", 100),
		col("punches", "Punches", width=220),
		col("attendance", "Attendance", "Link", 170, "Attendance"),
		col("status", "Status", width=90),
		col("working_hours", "Hours", "Float", 70),
		col("shift", "Shift", "Link", 130, "Shift Type"),
		col("verdict_label", "Verdict", width=260),
		col("detail", "Detail", width=420),
		col("repair", "Repair", width=80),
	]
